core/domain/user_query_jobs_one_off.py

```diff
@@ -0,0 +1,104 @@
+# coding: utf-8
+#
+# Copyright 2016 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs to execute and get result of a query."""
+import ast
+import datetime
+
+from core import jobs
+from core.domain import rights_manager
+from core.platform import models
+
+(user_models, exp_models) = (
+    models.Registry.import_models(
+        [models.NAMES.user, models.NAMES.exploration]))
+
+# pylint: disable=too-many-return-statements
+
+class UserQueryOneOffJob(jobs.BaseMapReduceJobManager):
+    """One-off job for excuting query with given query parameters.
+    For each user we check if he/she satisfies query criteria. If the user
+    satisfies the query criteria, then yield a tuple (query_id, user_id).
+    The reducer function stores all user_ids that satisfy the query in the
+    corresponding UserQueryModel."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(user_settings_model):
+        query_id = jobs.BaseMapReduceJobManager.get_mapper_param('query_id')
+        query_model = user_models.UserQueryModel.get(query_id)
+        user_id = user_settings_model.id
+        user_contributions = user_models.UserContributionsModel.get(user_id)
+
+        if (user_id == query_model.submitter_id or
+                rights_manager.Actor(user_id).is_moderator()):
+            return
+
+        if query_model.login_in_last_n_days is not None:
+            if user_settings_model.last_logged_in:
+                difference = (
+                    datetime.datetime.utcnow() -
+                    user_settings_model.last_logged_in).days
+                if difference > query_model.login_in_last_n_days:
+                    return
```

Comment:
Wait, actually -- you raise a good point. I was thinking about it more, and I don't see why we would want to email a user who's logged in recently. If anything, it should be the other way around.

I checked the original issue, and it was ambiguous as to where the direction is. I think that instead of login_in_last_n_days we should instead have has_not_logged_in_for_n_days. Is it possible to change to that? I'm very sorry for not catching it earlier...